draft-ietf-tls-tlsflags.xml

```diff
@@ -121,17 +121,21 @@
                 documents will have to define which bit to set to show support, and the order of the bits
                 within the bit string shall be enumerated in network order: bit zero is the high-order bit of
                 the first octet as the flags field is transmitted.</t>
-            <t> A client that supports this extension SHALL send this extension with the flags field having
-                bits set only for those extensions that it intends to set. If it does not wish to set any such
-                flags in the ClientHello message, it MAY send the extension empty (with length of zero), or it
-                may omit the extension altogether.</t> 
+        </section>
+        <section anchor="rules" title="Rules for The Flags Extension">
+            <t> A client that supports this extension and at least one flag extension SHALL send this extension 
+                with the flags field having bits set only for those extensions that it intends to set. It MUST 
+                NOT send this extension with a length of zero.</t> 
```

Comment:
You can't have all-zero values because the length of the FlagExtensions field is "the minimal length that allows it to encode all of the present flags", so it would be zero with no flags present.

Reply:
What action is mandated if this requirement is not met?

Reply:
I guess whatever you always do with a malformed extension. I don't think 8446 specifies this, but I think a ClientHello (just as an example) with a malformed extension is a malformed ClientHello which would result in dropping the connection.

Yes, this one is recoverable (by just ignoring the extension) and could be a candidate for Postel's law, but if the implementation is not coding this extension properly, how can you assume that the next extension starts where you think it starts?  It sort of makes sense if you assume that the generic header is encoded properly and so the boundaries are OK. But I don't think this extension deserves any special lenient handling.

Reply:
Yes, I would prefer it is this were clear: if there are trailing zero octets, it is bad and you should stop talking to the thing that generated that.

Reply:
Suggested text?

Reply:
Insert before "`<t>` Note that this document does not define any particular bits for this string. ":

> `<t>`An implementation that receives an all-zero value for this extension or a value that contains trailing zero bytes MUST generate a fatal illegal_parameter alert.`</t>`

Reply:
Updated